torch/_tensor_str.py

```diff
@@ -212,6 +191,57 @@ def _tensor_str_with_formatter(self, indent, formatter, summarize):
     tensor_str = (',' + '\n' * (dim - 1) + ' ' * (indent + 1)).join(slices)
     return '[' + tensor_str + ']'
 
+def _complex_vector_str(self, indent, real_formatter, imag_formatter, summarize):
+    # length includes spaces and comma between elements and the extra j for complex
+    element_length = 2 * real_formatter.width() + 3
+    elements_per_line = max(1, int(math.floor((PRINT_OPTS.linewidth - indent) / (element_length))))
+    char_per_line = element_length * elements_per_line
+
+    def _complex_scalar(val, real_formatter=real_formatter, imag_formatter=imag_formatter):
+        real_str = real_formatter.format(val.real)
+        imag_str = imag_formatter.format(val.imag) + "j"
+        if val.imag < 0:
+            return real_str + imag_str.lstrip()
+        else:
+            return real_str + "+" + imag_str.lstrip()
+
+    if summarize and self.size(0) > 2 * PRINT_OPTS.edgeitems:
+        data = ([_complex_scalar(val) for val in self[:PRINT_OPTS.edgeitems].tolist()] +
+                [' ...'] +
+                [_complex_scalar(val) for val in self[-PRINT_OPTS.edgeitems:].tolist()])
+    else:
+        data = [_complex_scalar(val) for val in self.tolist()]
+
+    data_lines = [data[i:i + elements_per_line] for i in range(0, len(data), elements_per_line)]
+    lines = [', '.join(line) for line in data_lines]
+    return '[' + (',' + '\n' + ' ' * (indent + 1)).join(lines) + ']'
+
+def _complex_tensor_str_with_formatter(self, indent, real_formatter, imag_formatter, summarize):
+    dim = self.dim()
+
+    if dim == 0:
+        real_str = _scalar_str(self.real, real_formatter)
+        imag_str = _scalar_str(self.imag, imag_formatter) + "j"
+        if self.imag < 0:
+            return real_str + imag_str.lstrip()
+        else:
+            return real_str + "+" + imag_str.lstrip()
+
+    if dim == 1:
+        return _complex_vector_str(self, indent, real_formatter, imag_formatter, summarize)
+
+    if summarize and self.size(0) > 2 * PRINT_OPTS.edgeitems:
```

Comment:
I worry about the redundancy (and maintainability) of this function vs. the existing `_tensor_str_with_formatter`. In fact, even `_vector_str ` and `_complex_vector_str` have a lot of similarities. Would it be possible to push the complex code into `_vector_str`, or to make `_tensor_str_with_formatter` polymorphic to avoid some of the code duplication?

Reply:
yeah you are right. I was thinking about it too. I updated the PR to reuse the existing code wherever there was a common logic between real and complex printing

Reply:
You're correct - my mistake, I misread the line where they're constructed.